examples/capture_circular.py

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/python3
+
+from picamera2.encoders.h264_encoder import *
+from picamera2.picamera2 import *
+import time
+import os
+
+lsize = (320, 240)
+picam2 = Picamera2()
+video_config = picam2.video_configuration(main={"size": (1280, 720), "format": "RGB888"},
+                                          lores={"size": lsize, "format": "YUV420"})
+picam2.configure(video_config)
+picam2.start_preview()
+encoder = H264Encoder(1000000)
+encoder.buffersize = 30 * 5
+picam2.encoder = encoder
+picam2.start()
+picam2.start_encoder()
+
+w, h = lsize
+prev = None
+encoding = False
+ltime = 0
+
+while True:
+    cur = picam2.capture_buffer("lores")
+    cur = cur[:w * h].reshape(h, w)
+    if prev is not None:
+        # Measure pixels differences between current and
+        # previous frame
+        mse = np.square(np.subtract(cur, prev)).mean()
+        if mse > 7:
+            if not encoding:
+                epoch = int(time.time())
+                encoder.dumpbuffer("{}-before.h264".format(epoch))
+                encoder.output = open("{}.h264".format(epoch), 'wb')
```

Comment:
I'm slightly feeling that we haven't separated "encoding" from "outputting" very cleanly here. I think this would be tidier if we hid all the "buffer dumping" and file handling in a custom output class. The functionality here seems fine, we just perhaps rearrange where some of it lives. I'd be tempted to:

1. Create a CircularOutput class. You create it with a file name, but it accumulates stuff in a circular memory buffer, and it has the "dump" function which writes everything to the file, For some applications, this class might be useful in its own right. It would also make for a reasonably simple demo of a "custom output".

2.  Then there's another custom output class, which incorporates the CircularOutput and file outputs together, but I can't think of a good name for it! When you create it, it makes a CircularOutput object which it holds and initially just passes it everything that arrives. It has a "start_saving" function which the application code here would call. This just sets a flag, and when the output function sees the flag is set, it dumps the CircularOutput and starts saving stuff to file for real. It could even re-open the same file for "appending" (or you could optimise that process by not actually closing the file in between).

What do you think?

Reply:
I think this was a very good point, just been working on separating out the output functionality, which makes it nicer since it doesn't require using /dev/null, I've added a recording boolean that decides when frames are written.